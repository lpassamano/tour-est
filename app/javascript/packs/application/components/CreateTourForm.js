import React, { Component } from "react";
import { navigate } from "@reach/router";
import PropTypes from "prop-types";

class CreateTourForm extends Component {
  static defaultProps = { navigate };
  state = {
    title: "",
    starting_point: "",
    directions: "",
    estimated_time: "",
    topic: ""
  };

  handleSubmit = async event => {
    event.preventDefault();
    const { id, cultural_center } = this.props.currentStaffUser;
    const {
      title,
      starting_point,
      directions,
      estimated_time,
      topic
    } = this.state;
    const result = await this.props.onCreateTour({
      tour: {
        title: this.state.title,
        staff_user_id: id,
        cultural_center_id: cultural_center.id,
        starting_point: starting_point,
        directions: directions,
        estimated_time: estimated_time,
        topic: topic
      }
    });

    if (result.ok) {
      this.props.navigate("/admin");
      this.setState({ title: "" });
    } else {
      console.error(result.data.error);
    }
  };

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  render() {
    // TODO: mark title field as required
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          name="title"
          type="text"
          id="title"
          value={this.state.title}
          onChange={this.handleChange}
        />
        <input
          name="starting_point"
          type="text"
          id="starting_point"
          value={this.state.starting_point}
          onChange={this.handleChange}
        />
        <input
          name="directions"
          type="text"
          id="directions"
          value={this.state.directions}
          onChange={this.handleChange}
        />
        <input
          name="estimated_time"
          type="text"
          id="estimated_time"
          value={this.state.estimated_time}
          onChange={this.handleChange}
        />
        <input
          name="topic"
          type="text"
          id="topic"
          value={this.state.topic}
          onChange={this.handleChange}
        />
        <button type="submit">Create Tour</button>
      </form>
    );
  }
}

CreateTourForm.propTypes = {
  navigate: PropTypes.func.isRequired,
  onCreateTour: PropTypes.func.isRequired,
  currentStaffUser: PropTypes.shape({
    id: PropTypes.number.isRequired,
    cultural_center: PropTypes.shape({
      id: PropTypes.number.isRequired
    }).isRequired
  })
};

export default CreateTourForm;
