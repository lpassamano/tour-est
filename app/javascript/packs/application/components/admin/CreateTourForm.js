import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TourFormFields from "./TourFormFields";
import * as tourActions from "../../redux/tours/actions";
import * as tourSelectors from "../../redux/tours/selectors";

export class CreateTourForm extends Component {
  handleSubmit = attributes => {
    this.props.onCreateTour({ tour: attributes });
  };

  render() {
    return <TourFormFields onSubmit={this.handleSubmit} />;
  }
}

CreateTourForm.propTypes = {
  onCreateTour: PropTypes.func.isRequired
};

const mapDispatchToProps = { onCreateTour: tourActions.createTour };
const enhance = connect(
  null,
  mapDispatchToProps
);

export default enhance(CreateTourForm);
