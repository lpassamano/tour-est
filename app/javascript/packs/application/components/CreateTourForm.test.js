import CreateTourForm from "./CreateTourForm";
import React from "react";
import { mount } from "enzyme";

describe("<CreateTourForm />", () => {
  const setup = ({ onCreateTour = jest.fn(), navigate = jest.fn() }) => {
    const currentStaffUser = { id: 1, cultural_center: { id: 1 } };
    const component = mount(
      <CreateTourForm
        onCreateTour={onCreateTour}
        navigate={navigate}
        currentStaffUser={currentStaffUser}
      />
    );
    return component;
  };

  test("onSubmit - when create button is clicked the current value of title is submitted", done => {
    const onCreateTour = jest.fn();
    onCreateTour.mockResolvedValue({ ok: true });
    const navigate = to => {
      expect(to).toEqual("/admin");
      done();
    };
    const component = setup({ onCreateTour, navigate });
    const event = { target: { name: "title", value: "Best Tour Ever!" } };
    component.find("input#title").simulate("change", event);
    component.find("form").simulate("submit", { preventDefault: () => null });
    expect(onCreateTour).toHaveBeenCalledWith({
      tour: {
        title: "Best Tour Ever!",
        staff_user_id: 1,
        cultural_center_id: 1
      }
    });
  });

  test("handleChange - when text is typed into the title field the state is updated", () => {
    const component = setup({});
    const event = { target: { name: "title", value: "Best Tour Ever!" } };
    component.find("input#title").simulate("change", event);
    expect(component.state("title")).toEqual("Best Tour Ever!");
  });

  test("render", () => {
    const component = setup({});
    expect(component).toMatchSnapshot();
  });
});
