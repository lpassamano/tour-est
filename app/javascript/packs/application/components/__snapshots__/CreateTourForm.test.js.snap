// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateTourForm /> render 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    htmlFor="title"
  >
    Title:
  </label>
  <input
    id="title"
    name="title"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="starting_point"
  >
    Starting Point:
  </label>
  <input
    id="starting_point"
    name="starting_point"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="directions"
  >
    Directions:
  </label>
  <input
    id="directions"
    name="directions"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="estimated_time"
  >
    Estimated Time:
  </label>
  <input
    id="estimated_time"
    name="estimated_time"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="description"
  >
    Description:
  </label>
  <input
    id="description"
    name="description"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <button
    type="submit"
  >
    Create Tour
  </button>
</form>
`;
